modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/AWS.scala

```diff
@@ -125,8 +128,39 @@ object AWS {
       }
     }
 
-    def readSqs(name: String, stop: Stream[F, Boolean]): Stream[F, Message[F, String]] =
-      SQS.readQueue(name, sqsVisibility.toSeconds.toInt, stop).map { case (msg, ack, extend) => Message(msg.body(), ack, extend) }
+
+    def readSqs(name: String, stop: Stream[F, Boolean]): Stream[F, String] = {
+      val awakePeriod: FiniteDuration = sqsVisibility - ExtendAllowance
+      SQS.readQueue(name, sqsVisibility.toSeconds.toInt, stop).flatMap { case (msg, ack, extend) =>
+        Stream.emit(msg.body())
+          .concurrently {
+            Stream.awakeEvery[F](awakePeriod).evalMap { _ =>
+              Logging[F].info(s"Approaching end of SQS message visibility. Extending visibility by $sqsVisibility.") *>
+              extend(sqsVisibility)
+            }
+            .handleErrorWith { t =>
+              Stream.eval(Logging[F].error(t)("Error extending SQS message visibility"))
```

Comment:
Am I right that `extend` and then `handleErrorWith` will always be invoked for every message? I.e. in a previous implementation we called `extend` only for a messages that were in-fly mode long enough and now it seems we're just extending a message until it's ack'ed already and then error-handing branch comes in to terminate the stream?

Reply:
No, the `concurrently` process gets cancelled (without error) after the downstream has finished processing this stream of a single message, and immediately before running the `onFinalizeCase` block.  The `handleErrorWith` block is only run if there really is an unexpected exception.

Reply:
It's kinda amazing how it works!  If you want to play around with it, I was running this test program to simulate the sqs stream and the retry stream.  I attach the log output.
[output.txt](https://github.com/snowplow/snowplow-rdb-loader/files/8797884/output.txt)


```
  def run(args: List[String]): IO[ExitCode] = {

    val retryStream =
      Stream.emit("retry1") ++
      Stream.emit("retry2") ++
      Stream.emit("retry3") ++
      Stream.emit("retry4") ++
      Stream.emit("retry5")

    val sqsStream = Stream.awakeDelay[IO](2.seconds)
      .zipWithIndex
      .map { case (_, i) => s"sqs$i" }
      .flatMap { msg =>
        Stream.eval(IO.delay(println(s"Emitting $msg")).as(msg))
          .concurrently {
            Stream.awakeEvery[IO](1.second).evalMap { _ =>
              IO.delay(println(s"Extending $msg"))
            }.drain
          }
          .onFinalizeCase {
            case ExitCase.Canceled =>
              IO.delay(println(s"Cancelled $msg"))
            case ExitCase.Completed =>
              IO.delay(println(s"Completed $msg"))
            case ExitCase.Error(_) =>
              IO.delay(println(s"Handled error $msg"))
          }
      }

    retryStream.merge(sqsStream)
      .evalTap(i => IO.delay(println(s"PROCESSING start $i")))
      .evalTap(_ => IO.sleep(10.seconds))
      .evalTap(i => IO.delay(println(s"PROCESSING end $i")))
      .flatMap { i =>
        if (i == "sqs4") Stream.raiseError[IO](new RuntimeException("boom!")) else Stream.emit(i)
      }
      .compile
      .drain
      .as(ExitCode.Success)
  }
```